monix-execution/shared/src/main/scala/monix/execution/misc/Local.scala

```diff
@@ -294,8 +338,21 @@ final class Local[A](default: () => A) {
   def bind[R](value: A)(f: => R): R = {
     val parent = Local.getContext()
     Local.setContext(parent.bind(key, Some(value)))
-    try f
-    finally Local.setContext(parent)
+
+    f match {
```

Comment:
Note that this does not have the semantic of the original code — `f` is a by-name parameter and you're evaluating it twice.

Also the first time you're evaluating `f`, it isn't in the context of a try/catch statement. This means that if `f` throws an error, you're not restoring the context.

Also this pattern matching here is not future proof. It isn't idiomatic Scala so to speak. If we really need to discriminate between data types like this, then we should use implicits.

Reply:
> Note that this does not have the semantic of the original code — f is a by-name parameter and you're evaluating it twice.
> Also the first time you're evaluating f, it isn't in the context of a try/catch statement. This means that if f throws an error, you're not restoring the context.

Right, I will fix it

> Also this pattern matching here is not future proof. It isn't idiomatic Scala so to speak. If we really need to discriminate between data types like this, then we should use implicits.

What do you mean by "future proof" ?